Add registry utility functions

riyaa14 opened the pull request:
**Description**

This PR fixes [#13668](https://github.com/meshery/meshery/issues/13668)

**Notes for Reviewers**
#### Files moved completely from `mesheryctl/pkg/utils` to `meshery/meshkit`:
1. model.go
2. relationship.go
3. component.go
4. spreadsheet_update.go

#### New files created in meshkit
5. helper.go : some functions from mesheryctl/pkg/utils/formatter.go and mesheryctl/pkg/utilsregistry.go 
6. error.go : some specific error codes from mesheryctl/pkg/utils/formatter.go 

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.
 

<!--
Thank you for contributing to Meshery! 

Contributing Conventions:

1. Include descriptive PR titles with [<component-name>] prepended.
2. Build and test your changes before submitting a PR. 
3. Sign your commits

By following the community's contribution conventions upfront, the review process will 
be accelerated and your PR merged more quickly.
-->

leecalcote commented on registry/component.go:
```diff
@@ -0,0 +1,447 @@
+	SVG_HEIGHT = 20
+)
+
+type ComponentCSV struct {
```
@riyaa14 this generated from schema using the openapi tool?

riyaa14 replied:
I didn't generate it but rather this file was copied completely from mesheryctl utils and shifted here in meshkit/registry

leecalcote reviewed (commented):
Thank you for placing these files under "registry" folder.

lekaf974 commented on registry/spreadsheet_update.go:
```diff
@@ -0,0 +1,349 @@
+func VerifyandUpdateSpreadsheet(cred string, wg *sync.WaitGroup, srv *sheets.Service, spreadsheetUpdateChan chan SpreadsheetData, sheetId, modelCSVFilePath, componentCSVFilePath string) {
+	defer wg.Done()
+
+	entriesToBeAddedInCompSheet := []*ComponentCSV{}
+	entriesToBeAddedInModelSheet := []*ModelCSV{}
```
Moving at top of for loop will allow to remove the needs of `Reset the list`

riyaa14 replied:
According to my understanding, we can't because resetting the lists is based on conditions.
In the following condition only one list is updated - 
![Screenshot 2025-02-26 at 2 43 37 AM](https://github.com/user-attachments/assets/b09164cc-fa4d-4193-91ab-46f6e75146fe)

we aren't updating the lists after every iteration, it is conditional

lekaf974 commented on registry/spreadsheet_update.go:
```diff
@@ -0,0 +1,349 @@
+		if err != nil {
+			Log.Error(err)
+		}
+		return
```
This looks not useful here

lekaf974 commented on registry/spreadsheet_update.go:
```diff
@@ -0,0 +1,349 @@
+	if len(entriesToBeAddedInModelSheet) > 0 {
+		err := updateModelsSheet(srv, cred, sheetId, entriesToBeAddedInModelSheet)
+		if err != nil {
+			Log.Error(err)
```
Just to understand, we continue even if the update fail right ?

riyaa14 replied:
yes, we log the error and continue with rest of the logic

lekaf974 commented on registry/spreadsheet_update.go:
```diff
@@ -0,0 +1,349 @@
+	}
+	return modelCSVFilePath, componentCSVFilePath, relationshipCSVFilePath, nil
+}
+func getCSVHeader(filePath string) (headers, secondRow []string, err error) {
```
The function name is misleading since it does not return only the csv header

riyaa14 replied:
We are reading the [meshery integrations](https://docs.google.com/spreadsheets/d/1DZHnzxYWOlJ69Oguz4LkRVTFM79kC2tuvdwizOJmeMw/edit?usp=sharing) spreadsheet in which first two rows are headers, not just row one

That's why this function is named getCSVHeader, because row 1 and 2 are headers in this sheet
![Screenshot 2025-02-26 at 2 29 17 AM](https://github.com/user-attachments/assets/6c003a39-9df8-4ccf-939d-3028549be748)

What do you think? shall we still name it something different?

lekaf974 replied:
Looking the spreadsheet, the real header look to be the second row only but if I am right test both in case someone use a copy were header is the first row right ?

riyaa14 replied:
Yes ✅️

lekaf974 commented on registry/spreadsheet_update.go:
```diff
@@ -0,0 +1,349 @@
+	for _, file := range files {
+		filePath := filepath.Join(csvDirectory, file.Name())
+		if !file.IsDir() && strings.HasSuffix(file.Name(), ".csv") {
+			headers, secondRow, err := getCSVHeader(filePath)
```
should we not evaluate error first ?

riyaa14 replied:
yes, makes complete sense, moving the error up

lekaf974 reviewed (changes requested):
Did a first pass and will continue soon

riyaa14 commented on registry/helpers.go:
```diff
@@ -0,0 +1,63 @@
+)
+
+var (
+	// global logger variable
+	Log logger.Handler
+	//global logger error variable
+	LogError logger.Handler
```
https://github.com/meshery/meshery/blob/0ac709a34b0001b48bad6ed61dfe851fbb0036a8/mesheryctl/pkg/utils/helpers.go#L259 - that logger is under `utils` package in `mesheryctl` and this logger is associated with `registry` package in `meshkit`

lekaf974 replied:
oh did not notice

riyaa14 commented:
@lekaf974 Thank you for the review!
just to inform, I haven't made any changes to functions that I have "added" here
These files: model.go, relationship.go, component.go, spreadsheet_update.go are copied as it is from mesheryctl utils to here in meshkit.
Thus, I might not be sure of the reasoning behind certain logic written.

lekaf974 commented on registry/error.go:
```diff
@@ -0,0 +1,79 @@
+
+var (
+	ErrMarshalStructToCSVCode     = "meshkit-11301"
+	ErrAppendToSheetCode          = "meshkit-11302"
+	ErrUpdateToSheetCode          = "meshkit-11303"
+	ErrFileReadCode               = "meshkit-11304"
+	ErrGeneratesModelCode         = "meshkit-11305"
+	ErrReadCSVRowCode             = "meshkit-11306"
+	ErrCSVFileNotFoundCode        = "meshkit-11307"
+	ErrUpdateComponentsCode       = "meshkit-11308"
+	ErrGeneratesComponentCode     = "meshkit-11309"
+	ErrUpdateRelationshipFileCode = "meshkit-11310"
```
Some of those error seems not to be specific to registry so wondering if they should not be moved to https://github.com/meshery/meshkit/blob/master/utils/error.go

riyaa14 replied:
ErrMarshalStructToCSV, ErrFileRead, ErrReadCSVRow, ErrCSVFileNotFound -> these look like the ones that should be moved. right?

souvikinator replied:
Please check if any of the errors already exists in the meshkit and use those instead

lekaf974 commented:
@riyaa14 took a look on other PR I am seeing that mesheryctl/internal/cli/root/model/generate.go was also changed due to this PR so changes related to models needs to be in models folder not i the registry one

riyaa14 commented:
@lekaf974 the model generate shares one function with registry -> `getCSV`
model generate command first generates model, component and relationship CSVs from given spreadsheet. Then registers the models using these 3 CSVs (this is same as model import)

I can move this function to some common utility folder but since the function is tailored to meshery integrations sheet which is closely related to registry, getCSV should be a part of registry package in meshkit.

or we can move it to a common folder by naming it something like - `getCSVFromModelGenerationSheet`. What would you suggest?

souvikinator reviewed (approved):
LGTM

riyaa14 commented:
@souvikinator @lekaf974 Thanks for review, feedback incorporated ✅
Changes done - 

Errors related
- Errors related to CSVs moved to a new file created under `utils/csv` as `errors.go`
- ErrReadFile - moved to files package, as it seemed relevant there
-  `ErrCSVFileNotFoundCode` - this isn't moved as if you notice, its error messages mention model and component csvs, and thus specific to registry, thus shouldn't be moved to a common csv package

Correct placement for Contains function in registry/helpers.go
- Moved to utils package as it is a general function
- Changed name to Locate because Contains was already present, moreover, Locate seemed a better name for this function.
- Updated references in meshery and meshkit

lekaf974 commented on files/error.go:
```diff
@@ -9,6 +9,18 @@ import (
 )
 
+var (
+	ErrFileReadCode = "meshkit-11304"
```
Can you move this with other variable declaration ?

riyaa14 replied:
If you see the whole file, the variable declarations are separate for different error categories. 

Eg:

Here the error messages are specific to meshery designs
https://github.com/meshery/meshkit/blob/46ebe93256e39432588fe6e9c85cb19a909d2c5d/files/error.go#L24 

Here they are specific to archives
https://github.com/meshery/meshkit/blob/46ebe93256e39432588fe6e9c85cb19a909d2c5d/files/error.go#L330

and since ErrFileReadCode, is more generic and not related to these, I separated it out for better readability.

lekaf974 commented on files/error.go:
```diff
@@ -9,6 +9,18 @@ import (
+	ErrFileReadCode = "meshkit-11304"
+)
+
+func ErrFileRead(err error) error {
```
Can you move this behind following variables declaration

riyaa14 replied:
https://github.com/meshery/meshkit/pull/688#discussion_r1975867345

lekaf974 commented on utils/utils.go:
```diff
@@ -261,6 +261,15 @@ func Contains[G []K, K comparable](slice G, ele K) bool {
 	return false
 }
 
+func Locate(key string, col []string) int {
```
Perhaps this could be renamed so it gives more details on the purpose o the function since it has become an utils function.

riyaa14 commented:
@lekaf974 changes done as suggested ✅

lekaf974 reviewed (approved):
LGTM
